python/pyarrow/_flight.pyx

```diff
@@ -458,19 +535,34 @@ cdef class FlightServerBase:
         cdef:
             PyFlightServerVtable vtable = PyFlightServerVtable()
             int c_port = port
+        vtable.list_flights = &_list_flights
         vtable.get_flight_info = &_get_flight_info
         vtable.do_put = &_do_put
         vtable.do_get = &_do_get
+        vtable.list_actions = &_list_actions
+        vtable.do_action = &_do_action
         self.server.reset(new PyFlightServer(self, vtable))
         with nogil:
             self.server.get().Run(c_port)
 
+    def list_flights(self, criteria):
```

Comment:
Design note: I hadn't thought about this before, but perhaps it would be better for those user-defined methods to live in a separate `FlightServerHandler` object (imagine having different implementations of Flight, or perhaps different underlying protocols). This can be changed later.